File: 0200-number-of-islands/0200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:

        ROWS, COLS = len(grid), len(grid[0])
        dir = [[1,0],[-1,0],[0,1],[0,-1]]
        ans = 0
        seen = set()
        def dfs(r, c):
            seen.add((r, c))
            for dr,dc in dir:
                row, col = r + dr, c + dc
                if 0 <= row < ROWS and 0 <= col < COLS and (row, col) not in seen and grid[row][col] == '1':
                    dfs(row, col)
        
        for r in range(ROWS):
            for c in range(COLS):
                if (r, c) not in seen and grid[r][c] == '1':
                    ans += 1
                    dfs(r, c)
        
        return ans

File: 0200-number-of-islands/0200-number-of-islands.cpp
class Solution {
public:
    int ROWS;
    int COLS;
    vector<vector<int>> dir = {{1, 0},{-1, 0},{0, 1},{0, -1}};
    vector<vector<char>> grid;

    void dfs(int r, int c, vector<vector<bool>> &visited) {
        visited[r][c] = true;
        for (auto d: dir) {
            int row = r + d[0];
            int col = c + d[1];
            if (row >= 0 && row < ROWS && col >= 0 && col < COLS && !visited[row][col] && grid[row][col] == '1') {
                dfs(row, col, visited);
            }
        }
    }

    int numIslands(vector<vector<char>>& grid) {
        this->grid = grid;
        ROWS = grid.size();
        COLS = grid[0].size();
        vector<vector<bool>> visited(ROWS, vector<bool>(COLS, false));
        int res = 0;
        for (int r = 0; r < ROWS; r++) {
            for (int c = 0; c < COLS; c++) {
                if (!visited[r][c] && grid[r][c] == '1') {
                    res++;
                    dfs(r, c, visited);
                }
            }
        }
        return res;
    }
};